nixos/modules/services/networking/yggdrasil.nix

```diff
@@ -0,0 +1,119 @@
+in {
+  options = with types; {
+    services.yggdrasil = {
+      enable = mkEnableOption "the yggdrasil system service";
+
+      configFile = mkOption {
+        type =  nullOr str;
+        default = null;
+        example = "/run/keys/yggdrasil.conf";
+        description = ''
+          A file which contains JSON or HJSON configuration for
+          yggdrasil.
+
+          This does not have to be a complete configuration, as
+          yggdrasil will use default values for anything which is
+          omitted.  If the encryption and signing keys are omitted,
+          yggdrasil will generate new ones each time the service is
+          started, resulting in a random IPv6 address on the yggdrasil
+          network each time.
+
+          If this option is set, <option>config</option> will be ignored.
+        '';
+      };
+
+      config = mkOption {
+        type = str;
+        default = "";
+        description = ''
+          JSON or HJSON configuration for yggdrasil.
+
+          Warning: this is stored in the WORLD-READABLE Nix store!
+          Therefore, it is not appropriate for private keys.  If you
+          provide a partial configuration and do not specify the keys,
+          yggdrasil will generate a new set each time the service is
+          started, creating a random IPv6 address on the yggdrasil
+          network each time.
+
+          If you wish to specify the keys, use
+          <option>configFile</option>, which takes precedence over
+          <option>config</option>.
+        '';
+      };
+
+      port = mkOption {
+        type = nullOr int;
+        default = null;
+        example = 54321;
+           description = ''
+          Firewall port to open in order to permit incoming peerings.
+          This should match the port specified in <code>Listen</code>
+          in the yggdrasil configuration.  It is not necessary to open
+          a port in order to make outgoing peerings.
+        '';
+      };
+
+      openMulticastPort = mkOption {
+        type = bool;
+        default = true;
+        description = ''
+          Whether to open the UDP port used for multicast peer discovery
+          on local networks.
+        '';
+      };
+
+      package = mkOption {
+        type = package;
+        default = pkgs.yggdrasil;
+        defaultText = "pkgs.yggdrasil";
+        description = "Yggdrasil package to use.";
+      };
+
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.yggdrasil.configFile = mkDefault (
+      if cfg.config != "" then
+        toString(pkgs.writeTextFile {
+          name = "yggdrasil-conf";
+          text = cfg.config;
+        })
+      else null);
+
+    systemd.services.yggdrasil = {
+      description = "Yggdrasil Network Service";
+      path = [ cfg.package ];
+      bindsTo = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        mkdir -p /var/lib/yggdrasil/
```

Comment:
there is `systemd.tmpfiles` which is now the preferred way to setup the working directory of services, you can see some implementations in this PR:
https://github.com/NixOS/nixpkgs/pull/56265/files

Reply:
also, maybe we want to run the service as another user than root? IMHO it would be preferable to run services as less privileged users.

Reply:
Both of those are very good suggestions and I'll work on incorporating them.

Reply:
In this case you want `StateDirectory` instead of `tmpfiles`. Only use `tmpfiles` when `StateDirectory`, `RuntimeDirectory`, `CacheDirectory`, and `LogsDirectory` aren't flexible enough.

Reply:
Would `DynamicUser` be sufficient for this module?

Reply:
I think so and I'm working on trying it out.  One snag which I have run into is that the admin utility `yggdrasilctl` has a hardcoded expectation to either find its configuration file in `/etc/yggdrasil.conf` or a unix socket at `/var/run/yggdrasil.sock` and it looks like `DynamicUser` only lets services have write access to subdirectories of `/etc` and `/run`.  Is that correct?  If so I will have to patch `yggdrasilctl` to change its default filenames.

Reply:
`/var/run` is a symlink to `/run`. I think most distros have adopted that.

You could symlink the configuration file to `/etc` via `environment.etc` and if your user has the proper permissions I believe that would work, or run a patch in the source code to update the socket directory.

I'd also like to mention there is no hard requirement to use `DynamicUser` user here. Just use it when it makes sense to.

Reply:
Thanks for the symlink suggestion.

I've been unable to get yggdrasil to run using `DynamicUser` or any user other than root, even after experimenting with `capsh` to try to give that user as many of root's capabilities as possible.  So it is still running as root, but with capabilities stripped and many of systemd's sandboxing features applied, including most which are suggested in #20186.

Reply:
Just figured out DynamicUser and pushed an update.